File: sysmon-android/app/src/main/java/com/sysmon/monitor/viewmodel/MonitorViewModel.kt
package com.sysmon.monitor.viewmodel

import android.app.Application
import android.content.Intent
import android.os.Build
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.sysmon.monitor.SysMonApp
import com.sysmon.monitor.data.model.SystemMetrics
import com.sysmon.monitor.data.websocket.WsState
import com.sysmon.monitor.service.SysMonForegroundService
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * ViewModel 只负责 UI 状态管理：
 *  · 直接读取 SysMonApp.wsClient 的 state/metrics（与 Service 共享同一实例）
 *  · 用户操作通过 startService + Action 转发给 SysMonForegroundService 执行
 *  · 不再持有重连循环，不依赖 Activity 生命周期
 */
class MonitorViewModel(application: Application) : AndroidViewModel(application) {

    private val app      = application as SysMonApp
    private val wsClient = app.wsClient
    private val urlRepo  = app.urlRepo

    // ── 直接桥接 wsClient 的状态（与 Service 共享同一实例，状态天然一致）────────
    val wsState: StateFlow<WsState>        = wsClient.state
    val metrics: StateFlow<SystemMetrics?> = wsClient.metrics

    // ── 历史数据 ───────────────────────────────────────────────────────────────
    private val _cpuHistory   = MutableStateFlow<List<Float>>(emptyList())
    val cpuHistory: StateFlow<List<Float>> = _cpuHistory.asStateFlow()

    private val _memHistory   = MutableStateFlow<List<Float>>(emptyList())
    val memHistory: StateFlow<List<Float>> = _memHistory.asStateFlow()

    private val _netRxHistory = MutableStateFlow<List<Double>>(emptyList())
    val netRxHistory: StateFlow<List<Double>> = _netRxHistory.asStateFlow()

    private val _netTxHistory = MutableStateFlow<List<Double>>(emptyList())
    val netTxHistory: StateFlow<List<Double>> = _netTxHistory.asStateFlow()

    // ── 当前 URL 和连接列表 ────────────────────────────────────────────────────
    private val _wsUrl = MutableStateFlow("")
    val wsUrl: StateFlow<String> = _wsUrl.asStateFlow()

    private val _connectedUrl = MutableStateFlow("")
    val connectedUrl: StateFlow<String> = _connectedUrl.asStateFlow()

    val savedUrls: StateFlow<List<String>>    = urlRepo.urls
    val savedRemarks: StateFlow<List<String>> = urlRepo.remarks

    // cookie：全局 WebSocket 握手时附加，持久化在 UrlRepository
    val cookie: StateFlow<String> = urlRepo.cookie

    // autoConnecting：Connecting 状态时为 true，供 UI 显示"连接中"
    private val _autoConnecting = MutableStateFlow(false)
    val autoConnecting: StateFlow<Boolean> = _autoConnecting.asStateFlow()

    companion object {
        private const val TAG = "SysMonVM"
    }

    init {
        // 监听指标，追加历史
        viewModelScope.launch {
            wsClient.metrics.collect { m ->
                m ?: return@collect
                _cpuHistory.value   = (_cpuHistory.value   + m.cpuUsagePercent).takeLast(60)
                _memHistory.value   = (_memHistory.value   + m.memoryUsagePercent).takeLast(60)
                _netRxHistory.value = (_netRxHistory.value + m.netRxKbps).takeLast(60)
                _netTxHistory.value = (_netTxHistory.value + m.netTxKbps).takeLast(60)
            }
        }

        // 监听连接状态，更新 UI 标志
        viewModelScope.launch {
            wsClient.state.collect { state ->
                Log.d(TAG, "UI state: $state")
                when (state) {
                    is WsState.Connected   -> {
                        _connectedUrl.value = _wsUrl.value
                        _autoConnecting.value = false
                    }
                    is WsState.Connecting  -> _autoConnecting.value = true
                    is WsState.Disconnected,
                    is WsState.Error       -> _autoConnecting.value = false
                    else                   -> {}
                }
            }
        }

        // 启动前台 Service（Service 负责实际连接和重连循环）
        ensureServiceRunning()
    }

    // ── 公开操作 ───────────────────────────────────────────────────────────────

    fun updateUrl(url: String) { _wsUrl.value = url }

    /** 用户手动点击连接 */
    fun connect() {
        val url = _wsUrl.value
        sendToService(SysMonForegroundService.ACTION_CONNECT, url, urlRepo.cookie.value)
    }

    /** 用户手动断开，不再自动重连 */
    fun disconnect() {
        sendToService(SysMonForegroundService.ACTION_DISCONNECT, null)
        clearHistory()
    }

    /** 取消自动连接（不标记 manually，后续可继续） */
    fun cancelAutoConnect() {
        sendToService(SysMonForegroundService.ACTION_CANCEL, null)
    }

    /** 切换到指定 URL（静默换连，不触发熄屏逻辑） */
    fun connectTo(url: String) {
        _wsUrl.value = url
        _connectedUrl.value = url
        clearHistory()
        sendToService(SysMonForegroundService.ACTION_RECONNECT, url, urlRepo.cookie.value)
    }

    /** 保存/更新 Cookie */
    fun saveCookie(cookie: String) { urlRepo.saveCookie(cookie) }

    /**
     * onResume 时调用。
     * Service 已在后台自主管理重连循环，这里只做一次轻量检查：
     * 若 Service 不在运行则重新启动（异常重启恢复场景）。
     */
    fun retryAutoConnect() {
        ensureServiceRunning()
    }

    fun saveCurrentUrl() {
        val url = _wsUrl.value.trim()
        if (url.isNotEmpty()) urlRepo.addUrl(url)
    }

    fun removeUrl(url: String) { urlRepo.removeUrl(url) }

    fun switchToPrevUrl() {
        val urls = urlRepo.urls.value
        if (urls.size < 2) return
        val baseUrl = currentSwitchBaseUrl(urls)
        val idx = urls.indexOf(baseUrl)
        val target = urls[if (idx <= 0) urls.lastIndex else idx - 1]
        connectTo(target)
    }

    fun switchToNextUrl() {
        val urls = urlRepo.urls.value
        if (urls.size < 2) return
        val baseUrl = currentSwitchBaseUrl(urls)
        val idx = urls.indexOf(baseUrl)
        val target = urls[if (idx < 0 || idx >= urls.lastIndex) 0 else idx + 1]
        connectTo(target)
    }

    fun saveRemark(url: String, remark: String) { urlRepo.saveRemark(url, remark) }
    fun getRemarkFor(url: String): String = urlRepo.getRemarkFor(url)

    // ── 内部工具 ───────────────────────────────────────────────────────────────


    private fun currentSwitchBaseUrl(urls: List<String>): String {
        val pendingUrl = _wsUrl.value
        if (pendingUrl in urls) return pendingUrl

        val connectedUrl = _connectedUrl.value
        if (connectedUrl in urls) return connectedUrl

        return urls.first()
    }

    private fun ensureServiceRunning() {
        val ctx = getApplication<Application>()
        val intent = SysMonForegroundService.startIntent(ctx)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            ctx.startForegroundService(intent)
        } else {
            ctx.startService(intent)
        }
    }

    private fun sendToService(action: String, url: String?, cookie: String = "") {
        val ctx = getApplication<Application>()
        val intent = Intent(ctx, SysMonForegroundService::class.java).apply {
            this.action = action
            url?.let { putExtra(SysMonForegroundService.EXTRA_URL, it) }
            if (cookie.isNotEmpty()) putExtra(SysMonForegroundService.EXTRA_COOKIE, cookie)
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            ctx.startForegroundService(intent)
        } else {
            ctx.startService(intent)
        }
    }

    private fun clearHistory() {
        _cpuHistory.value   = emptyList()
        _memHistory.value   = emptyList()
        _netRxHistory.value = emptyList()
        _netTxHistory.value = emptyList()
    }

    override fun onCleared() {
        super.onCleared()
        // ViewModel 销毁时不断开连接，连接由 Service 持有
    }
}

File: sysmon-ios/SysMon/Sources/ViewModel/MonitorViewModel.swift
import Foundation
import Observation

// ══════════════════════════════════════════════════════════════════════════════
// MonitorViewModel  —  使用 Swift 5.9 @Observable 宏（iOS 17+）
// 主线程绑定：所有属性直接驱动 SwiftUI 视图更新
// ══════════════════════════════════════════════════════════════════════════════

@Observable
@MainActor
final class MonitorViewModel {

    // ── 连接状态 ────────────────────────────────────────────────────────────
    var wsState:      WsState        = .disconnected
    var wsUrl:        String         = "ws://192.168.1.100:9001"
    var connectedUrl: String         = ""
    var autoConnecting: Bool         = false

    // ── 实时指标 ────────────────────────────────────────────────────────────
    var metrics: SystemMetrics?

    // ── 历史数据（60 个点）────────────────────────────────────────────────
    var cpuHistory:   [Float]  = []
    var memHistory:   [Float]  = []
    var netRxHistory: [Double] = []
    var netTxHistory: [Double] = []

    // ── 已保存的连接（存储属性，让 @Observable 能追踪变化）────────────────────
    var savedUrls:    [String] = []
    var savedRemarks: [String] = []

    // ── 内部成员 ─────────────────────────────────────────────────────────
    private let wsClient  = SysMonWebSocket()
    private let urlRepo   = UrlRepository.shared

    private var manuallyDisconnected = false
    private var autoConnectTask:  Task<Void, Never>?
    private var reconnectTask:    Task<Void, Never>?

    private static let maxHistory:         Int          = 60
    private static let autoConnectTimeout: Duration     = .seconds(5)
    private static let reconnectDelay:     Duration     = .seconds(8)

    // ── 初始化 ──────────────────────────────────────────────────────────────

    init() {
        // 从仓库加载初始数据
        savedUrls    = UrlRepository.shared.urls
        savedRemarks = UrlRepository.shared.remarks
        bindWebSocket()
        Task { await tryAutoConnect() }
    }

    // ── WebSocket 回调绑定 ──────────────────────────────────────────────────

    private func bindWebSocket() {
        wsClient.onConnected = { [weak self] in
            guard let self else { return }
            self.wsState      = .connected
            self.connectedUrl = self.wsUrl
            self.saveCurrentUrl()
            self.stopAutoConnect()
        }

        wsClient.onDisconnected = { [weak self] reason in
            guard let self else { return }
            let wasConnected = (self.wsState == .connected)
            self.wsState = .disconnected
            if wasConnected, !self.manuallyDisconnected {
                self.scheduleReconnect()
            }
        }

        wsClient.onError = { [weak self] message in
            guard let self else { return }
            self.wsState = .error(message)
            if !self.manuallyDisconnected {
                self.scheduleReconnect()
            }
        }

        wsClient.onMetrics = { [weak self] m in
            guard let self else { return }
            self.metrics = m
            self.appendHistory(m)
        }
    }

    // ── 自动连接（逐个尝试已保存的 URL）───────────────────────────────────────

    private func tryAutoConnect() async {
        let urls = urlRepo.urls
        guard !urls.isEmpty, wsState != .connected else { return }

        stopAutoConnect()
        autoConnecting = true

        autoConnectTask = Task { [weak self] in
            guard let self else { return }
            for url in urls {
                guard !Task.isCancelled else { break }
                if self.wsState == .connected { break }

                await MainActor.run {
                    self.wsUrl = url
                    self.wsClient.connect(urlString: url)
                }

                // 等待连接结果（最多 5 秒）
                let deadline = Date().addingTimeInterval(5)
                while Date() < deadline && !Task.isCancelled {
                    let currentState = self.wsState
                    if currentState == .connected { break }
                    if case .error = currentState { break }
                    if case .disconnected = currentState, currentState != .connecting { break }
                    try? await Task.sleep(for: .milliseconds(200))
                }

                if self.wsState == .connected { break }

                await MainActor.run { self.wsClient.disconnect() }
                try? await Task.sleep(for: .milliseconds(300))
            }

            await MainActor.run { [weak self] in
                self?.autoConnecting = false
            }
        }
        await autoConnectTask?.value
        autoConnecting = false
    }

    private func stopAutoConnect() {
        autoConnectTask?.cancel()
        autoConnectTask = nil
        autoConnecting  = false
    }

    private func scheduleReconnect() {
        guard !manuallyDisconnected, !urlRepo.urls.isEmpty else { return }
        reconnectTask?.cancel()
        reconnectTask = Task { [weak self] in
            try? await Task.sleep(for: Self.reconnectDelay)
            guard !Task.isCancelled else { return }
            await self?.tryAutoConnect()
        }
    }

    // ── 公开操作 ─────────────────────────────────────────────────────────────

    func connect() {
        manuallyDisconnected = false
        stopAutoConnect()
        reconnectTask?.cancel()
        wsClient.connect(urlString: wsUrl)
    }

    func disconnect() {
        manuallyDisconnected = true
        stopAutoConnect()
        reconnectTask?.cancel()
        wsClient.disconnect()
        clearHistory()
    }

    func connectTo(_ url: String) {
        manuallyDisconnected = false
        stopAutoConnect()
        reconnectTask?.cancel()
        wsUrl = url
        wsClient.disconnect()
        clearHistory()
        wsClient.connect(urlString: url)
    }

    func switchToPrevUrl() {
        let urls = urlRepo.urls
        guard urls.count >= 2 else { return }
        let idx    = urls.firstIndex(of: connectedUrl) ?? 0
        let target = urls[idx <= 0 ? urls.count - 1 : idx - 1]
        connectTo(target)
    }

    func switchToNextUrl() {
        let urls = urlRepo.urls
        guard urls.count >= 2 else { return }
        let idx    = urls.firstIndex(of: connectedUrl) ?? -1
        let target = urls[(idx < 0 || idx >= urls.count - 1) ? 0 : idx + 1]
        connectTo(target)
    }

    func saveCurrentUrl() {
        let url = wsUrl.trimmingCharacters(in: .whitespaces)
        guard !url.isEmpty else { return }
        urlRepo.addUrl(url)
        syncSavedUrls()
    }

    func removeUrl(_ url: String) {
        urlRepo.removeUrl(url)
        syncSavedUrls()
    }

    func saveRemark(_ remark: String, for url: String) {
        urlRepo.saveRemark(remark, for: url)
        syncSavedUrls()
    }

    private func syncSavedUrls() {
        savedUrls    = urlRepo.urls
        savedRemarks = urlRepo.remarks
    }

    func getRemark(for url: String) -> String {
        urlRepo.getRemark(for: url)
    }

    // ── 内部工具 ─────────────────────────────────────────────────────────────

    private func appendHistory(_ m: SystemMetrics) {
        func append<T>(_ arr: inout [T], _ val: T) {
            arr.append(val)
            if arr.count > Self.maxHistory { arr.removeFirst() }
        }
        append(&cpuHistory,   m.cpuUsagePercent)
        append(&memHistory,   m.memoryUsagePercent)
        append(&netRxHistory, m.netRxKbps)
        append(&netTxHistory, m.netTxKbps)
    }

    private func clearHistory() {
        metrics      = nil
        cpuHistory   = []
        memHistory   = []
        netRxHistory = []
        netTxHistory = []
    }
}
